### DISCUSSION TITLE
Scoping, typings, references, autocompletion structure guidance

### DISCUSSION THREAD

**Discussion by kristiannotari** [eclipse-langium/langium#856] [Q&A]
Hello everyone,

I've already received a lot of help for my DSL regarding builtin library [here](https://github.com/langium/langium/discussions/853).

I'm new to both DSL creation and Langium and I'd like to better understand the document lifecycle, and how I should deal with scoping, typing, references and such. I've read the [document lifecycle](https://langium.org/docs/document-lifecycle/) guide but I'm missing how to properly structure my code given my DSL constraints.

I'd like not to bother you all by creating a discussion uniquely to guide me creating my DSL, but I wanted to get a clearer idea on how to structure things in Langium.

Given this DSL example file (limited example, pretty straightforward to read):
```
function f (arg1: number, arg2: string)

entity A {
    a1 = f(42, "hello") 
}

entity B {
   b1 = f(42, "hello from b")
}

entity A {
   a2 = f(43, "world")
}

compare a1, b1
```

and given these requirements:

1. Each file is isolated with its own definitions, so no references to other files
2. I want to keep track of function definitions (file scoped, no nesting, accessible to every point in the file)
3. I want to track entity definitions which can be repeated later in the same file (and refer to the same entity) and the named values (`a1`, `a2`, etc) inside them (each entity can refer to its own named values only known at a given point in the file, so for example the second entity block of `A` can access the `a1` named value defined earlier)
4. `compare` can access every named value from every entity, ignoring entity scoping
5. I want to create a type system for function definitions so that their application can be checked

Solutions/structures/questions I see/have:

1. global export for each file is empty (so no cross document referencing can happen) --> `ScopeComputation.computeExports` returns an empty array
2. functions definitions should be put in a "file global scope" so: `ScopeComputation.computeLocalScopes` should create a `PrecomputedScopes` with a bunch of function definitions ASTDescriptions with the "AST root node" as the containing node and then in the `ScopeProvider.getScope` how can I get those functions definitions? In the `DefaultScopeProvider` there's a "global scope" created with the `indexManager.allElements` method, but I can't understand how's that working since it's filtering with the `ReferenceType`???
3. entity definitions in the `ScopeComputation.computeLocalScopes` should be scoped under each `entity` definition block node, adding all the named values inside them. Doing so I create N scopes where N is the amount of `entity` definition blocks I encounter, even if repeated for the same entity. How can I then implement in the `ScopeProvider` the ability to access every named value know UP to the point I am? (example: in the first entity `A` definition block I should not see `a2` since it's defined later)
4. in the `ScopeProvider`, when retrieving the scope for a `compare` node I should get every named value (and function) basically. Should I retrieve the "file global scope" (see point 2) and add to that each precomputed scope from each entity definition block in the entire document/file, going upwards in the ast node from where I am?
5. This is almost unknown territory, since I've read there's no "default" implementation for a type system and one has to come with one himself. I can follow along what's has been done so far (looking at the [lox example](https://github.com/langium/langium-lox/tree/main/src/language-server) or similar) but what's worrying me is where to put "validation/checks" since I need some precomputed stuff (function definitions typings for example) in advance to actually compute if a function application is well-typed. Is the "validator" the right place to add these checks?

I'd like to know if my understanding is correct or if I'm mistaken somewhere. Thanks in advance!

EDIT: bonus question: is there some way to visually grasp what's being built by the various "passes" from text to scopes and stuff? Like some diagrams populating based on cursor position in a DSL file or something like that? Cause that would help greatly to see what's being built behind the scene. Or is there another way to properly debug these "internals" without having to rely on editor hints to know what's available in scope?

**Comment by msujew** [comment]
Hey @kristiannotari,

1. Yep, that seems reasonable
2. 

> how can I get those functions definitions?

You can always do it like the [linker does](https://github.com/langium/langium/blob/c5d912ba4ad35c2494f4598b1c627ea511782cb3/packages/langium/src/references/linker.ts#L199-L203). 

> but I can't understand how's that working since it's filtering with the ReferenceType???

Since every cross reference has a special cross reference type, we don't want to have all objects in scope which are visible in the index, but only those which fit the cross reference type. That's why we apply a filtering there.

3.

> How can I then implement in the ScopeProvider the ability to access every named value know UP to the point I am?

This is kind of a weird requirement, since normal language don't work that way. If a nested element is visible to a certain scope (i.e. in your case the root scope which contains the `compare` statement), it's also visible to all elements within that scope. I'm sure you could filter these elements in the `getScope` method, but it seems semantically unsound to design a language that way.

4.

> Should I retrieve the "file global scope" (see point 2) and add to that each precomputed scope from each entity definition block in the entire document/file, going upwards in the ast node from where I am?

Yes, that's what I would have recommended 👍 

5.

> in advance to actually compute if a function application is well-typed. Is the "validator" the right place to add these checks?

In advance compared to what? A code generator? Yes, the validator is usually always the right place to add these checks when you have a language which could contain semantic errors. There's no other intended way to make users aware of semantic errors in their documents other than the validator infrastructure. The Lox language implementation does the same.

Bonus Question:

Not really, or at least nothing that Langium provides as a builtin feature. To debug the scoping at a certain position in the document, I usually just run the code completion via `Ctrl+Space`. For everything else, live debugging is a bit more complicated. I recommend building a test suite for your language, so you can test parts of your language in isolation.

**Reply by kristiannotari** [reply]
1. ok
2. so basically the linker method you've *linked* (pun intended) goes from an ASTDescription to a node (from which he can do whatever he wants), so that would be how to retrieve the actual node. I meant how to retrieve the ASTDescriptions of the function definitions from within the `ScopeProvider`. The reference type thing is useful only if you use cross references at a grammar level with `value=[Rule:ID]` for example? If I don't use those I can disable filtering?
3. I'm basically recreating typescript interfaces with properties as values instead of types, so they have to be defined first (top-bottom in the file) before they can actually be referenced (examples as typescript interfaces):

✅ ok
<img width="555" alt="image" src="https://user-images.githubusercontent.com/17885006/210064653-b3f7dd9e-3bdf-40aa-a17b-698a0d9a9123.png">
❌ wrong, shall not use `b` before declaration
<img width="562" alt="image" src="https://user-images.githubusercontent.com/17885006/210064875-32cb0998-da36-4a7b-8ca0-ab3f9fd5c0e1.png">

4. Ok but I have a bonus question for this fourth point (😄): since I'm retrieving all the named values from all the entity definition blocks, they can be referenced with their fully qualified name (`A.a1` for example) as well as directly with the property name `a1` (so I assume I should precompute their scopes with 2 different descriptions with both name options). Of course, one can potentially write the same named value in another entity (and I should allow this) so, ONLY WHEN THIS OCCURS, I should warn by saying *"possible ambiguity, you're referring to `a1` which is defined in both `A` and `B` entities, you should use its fully qualified name to reference it `A.a1` or `B.a1`"*. Where/how to put this check? The more I progress structuring the DSL the more I feel the need for some additional custom map/dictionary/set of values to keep track of things like this but I'm not sure if that's the case (and how to keep track of these values within different classes and accessing them correctly, on a per document basis).

5. Yeah I'm not speaking about generation. I write my DSL file and I want it to be validated (both syntactically and semantically, with some custom domain specific validations) and type checked. So the validator is the right place to put all these kind of checks right? When I need type checking and complex semantic checks I can potentially use (and you're all are advising to do so) additional data structure to keep track of things I need to validate parts of the document, right? I mean, that's a thing we can do, right?

Bonus question:

I meant it as a tool to incrementally try out things, not like properly testing stuff. Just so that you could visualize what's being built while you're editing it, so you can try out different things and different structures and see scopes, indexes, types system (caches), and stuff like that.

**Reply by msujew** [reply]
> I meant how to retrieve the ASTDescriptions of the function definitions from within the ScopeProvider.

For creating descriptions from objects, you can always call the `AstNodeDescriptionProvider` or the [`createScopeForNodes`](https://github.com/langium/langium/blob/c5d912ba4ad35c2494f4598b1c627ea511782cb3/packages/langium/src/references/scope-provider.ts#L152-L161) function directly.

>  The reference type thing is useful only if you use cross references at a grammar level with value=[Rule:ID] for example? If I don't use those I can disable filtering?

The filtering is in theory only done as a default mechanism, so you could disable that, yes. Note that I wouldn't recommend doing so, as there are better ways of dealing with that. Most notably just creating a union type which collects all possible types you want to reference. Like `type Declaration = Function | Variable | Class | ...;` and then use that `value=[Declaration:ID]`.

> I'm basically recreating typescript interfaces with properties as values instead of types, so they have to be defined first (top-bottom in the file) before they can actually be referenced (examples as typescript interfaces):

Alright, I mean you can always go through the syntax tree of your document and eliminate all elements in a scope which appear after the container element of the to-be-scoped element.

> Where/how to put this check? The more I progress structuring the DSL the more I feel the need for some additional custom map/dictionary/set of values to keep track of things like this but I'm not sure if that's the case (and how to keep track of these values within different classes and accessing them correctly, on a per document basis).

Yeah, this is likely the point where you perform a lot of duplicate computations if you do everything in the validation phase. You can append a listener to document builder phases (similar to how we do it [here](https://github.com/langium/langium/blob/main/packages/langium/src/grammar/langium-grammar-module.ts#L84) for our grammar DSL). You probably want to have your preprocessing phase after the `ComputedScopes` phase, where you can start resolving cross references and so on to populate your document state. I guess that also answers the 5. point. Note that I would recommend to [extend the `LangiumDocument` interface](https://github.com/langium/langium/blob/c5d912ba4ad35c2494f4598b1c627ea511782cb3/packages/langium/src/grammar/workspace/documents.ts#L15-L17) on your part to store the additional information. The document lifecycle will automatically clean up any leftover data after a document change, so you don't need to worry about that.

> I meant it as a tool to incrementally try out things, not like properly testing stuff. Just so that you could visualize what's being built while you're editing it, so you can try out different things and different structures and see scopes, indexes, types system (caches), and stuff like that.

I see. No we hadn't put any thought so far into in-depth dev-tooling for Langium.

**Reply by kristiannotari** [reply]
Ok sorry for the recap but I'm starting to doubt my understanding of scopes and references in the context of Langium. My understanding:

1. cross-references (the ones put in the grammar as `[Declaration:ID]`) are there to ease the validation of referencing entities from a point in a file to another (potentially cross-file) without having to check if the referenced entity is correctly typed, since you're doing that with the `Declaration` bit of the cross reference directly in the grammar rule with [Declaration:ID], hence the filtering. Does this alone add some validation to the document or is that just for auto completion purposes / metadata so that you can do some work later?
2. scopes are there just as metadata for autocompletion and do not add any validation to the parsed document. The validation is something you need to do yourself, for example to check if "I can access this class member here" in a class-full programming language.
3. I don't see how well (given my understanding of points 1 and 2), langium different steps in the document lifecycle can help me better build and check DSL files (after the parsing step obviously). What if I get the parsed AST and just build custom validations and structures from them? What the built-in steps (and so extending default classes) helps me out doing? I can't properly picture how it's useful for me to have those steps/phases when I build up scopes for example (may highly depend on this being my first time actually trying to make a DSL works correctly and be validated in an editor-like environment).

EDIT: bonus question: are there some specific properties for rule Langium uses by default, like "name" for rules and so cross-references?

EDIT 2: bonus question 2: I have entities block which can define multiple entity names in a single block, therefore when I want to use cross-references my "name" property inside the entity block definition rule is an array of names (ID) and not a single one. Is that something problematic?

**Reply by msujew** [reply]
> cross-references (the ones put in the grammar as [Declaration:ID]) are there to ease the validation of referencing entities from a point in a file to another (potentially cross-file) without having to check if the referenced entity is correctly typed, since you're doing that with the Declaration bit of the cross reference directly in the grammar rule with [Declaration:ID], hence the filtering. Does this alone add some validation to the document or is that just for auto completion purposes / metadata so that you can do some work later?

It's less about the referenced element being correctly typed, but more about what you want the referencable elements to be. Let me explain this using an example from TypeScript:

```ts
type X {
    name: string
}
const X = {
  name: 'X'
}
const x: X // references type X
const y = X // references const X
```

Both of these elements are named `X`, even though they live on the same block level. This is valid, since TypeScript has two different kinds of scopes, one for the type level and one for the runtime level. You can imagine this in Langium as one reference being `[Type:ID]` and the other being `[Variable:ID]`. In this case we want the language to scope to specific elements of a specific type depending on the context. And yes, this also helps the framework make educated guesses for the purpose of auto completion.

> scopes are there just as metadata for autocompletion and do not add any validation to the parsed document. The validation is something you need to do yourself, for example to check if "I can access this class member here" in a class-full programming language.

Nope, you indeed seem to have an incorrect conception of what the scoping mechanism is. Scopes contain descriptions which point to real elements in your file/workspace. They are used to resolve the [`Reference#ref`](https://github.com/langium/langium/blob/c5d912ba4ad35c2494f4598b1c627ea511782cb3/packages/langium/src/syntax-tree.ts#L50-L56) value in your AST. Through that you can traverse your AST using real references to other elements in your file, not only IDs. Consequently, Langium also automatically produces error messages for the user in case a reference cannot be resolved. See this example in our arithmetics language:

![image](https://user-images.githubusercontent.com/4377073/210140762-cfbe0722-144a-46f6-a0a0-e5b9a617b374.png)

It also uses these references to build a few language services on top, such as going to declarations, finding references, and so on. Using it as a means for auto completion is just a convenience feature, as both the referencing system and the completion system are able to stay in sync that way. Users usually only want to see those elements in there, which are actually referencable. All others automatically result in a linking error anyway.

> I don't see how well (given my understanding of points 1 and 2), langium different steps in the document lifecycle can help me better build and check DSL files (after the parsing step obviously). What if I get the parsed AST and just build custom validations and structures from them? What the built-in steps (and so extending default classes) helps me out doing? I can't properly picture how it's useful for me to have those steps/phases when I build up scopes for example (may highly depend on this being my first time actually trying to make a DSL works correctly and be validated in an editor-like environment).

As mentioned, the scoping/linking part of Langium is way more powerful than you think it is. Note that most users don't need to dive as deep into the document lifecycle to build their languages, as most languages consist of the grammar + scoping + validation part, which are cleanly separated into their own services. Langium handles the infrastructure for you, you just need to add the semantics of your language.

>  are there some specific properties for rule Langium uses by default, like "name" for rules and so cross-references?

Yes, it will use the property named `name`. Though you can modify that using the `NameProvider` service.

> I have entities block which can define multiple entity names in a single block, therefore when I want to use cross-references my "name" property inside the entity block definition rule is an array of names (ID) and not a single one. Is that something problematic?

Yes, that will be problematic. The best practice solution for this to override the name provider and return a single canonical name for the element. Then you also override the scope computation to basically create multiple descriptions for elements of this type, one for each name. Each name will basically be an alias for the actual element.

**Reply by kristiannotari** [reply]
Thanks for the answer, this helps a lot. Ok so I got something working by defining scopes in the computation step (no scope provider implementation yet) and everything seems to work out of the box: go to definition for things I specified in my builtin with the editor showing up my builtin library as a separate file (that's great!). I tried cross references so it added automatically autocompletion for what I am supposed to put in some places in my DSL and it's working great. Also, go to definition properly works for things I defined in the same DSL file (for the things I added scope computation for).

For the problematic bit:

- I understood I should create multiple descriptions and that's what I implemented right now
- I understood the name provider override, so that I'm basically "linking" every entity to the block where I've defined them plus other ones. The problem being, I have multiple entity definition possibly for the same entity, but with potentially different sets of entities each time. I have the "typescript interface" redefinition problem PLUS the fact (following typescript example) I can define multiple interface names at the same time as in (custom typescript with multiple interface names definitions):

```typescript
interface A, B { // <-- name provider "A" (keeping the first one), descriptions for A, B
   ab: "hello"
}
interface A, C { // <-- name provider "A", descriptions for A, C
   ac: "world"
}
interface B, C { // <-- name provider "B", descriptions for B, C
   bc: 42
}
```

Using `B` somewhere should link me to the first definition for B, same for everything else. Could another solution be to have the name provider returning a somehow concatenated string of names where I use a special separator (from a character which can not be an ID character) and then also override the cross-reference implementation to retrieve the nodes where the name includes (and not equals) the entity name? 

Or is there some "magical" way to have added AST nodes which don't reflect actual text parts of the file but are "generated" on the fly by some desugaring of expressions, like defining the same things for N entities resulting in an ordered list of N definitions in the AST?

**Reply by msujew** [reply]
What you're looking for could either be implemented by something like [a qualified name](https://langium.org/guides/scoping/qualified-name/) or a [member call scoping algorithm](https://langium.org/guides/scoping/class-member/). Qualified names are fairly simple to implement, since they purely rely on the scope computation and names of container elements. Member style scoping is more complicated, as it requires some sort of type system in place to infer class/interface information. Both can accomplish in theory the same, with the member style being more powerful in general.

For implementing qualified name scoping, you will need a qualified name data type rule, something like `QualifiedName returns string: ID ('.' ID)*`. This circumvents the lexing problem, since the qualified name is a parser rule (although it merely returns a string and not an AST node).

**Reply by kristiannotari** [reply]
Not sure I've understood how following qualified names and member call scoping techniques could help me with the multiple entity definition in a single block but I think I managed to get it working by simply computing scopes for each entity and then when I use the entity somewhere in the file it gets a nice clickable hover "go to definition"-like behavior which tells me everything's working correctly I guess, since the reference is resolved comparing the name I give in the description and not the name retrieved by the name provider for the scoped node (solving the "multiple names in one block" problem).

Now I have another problem which is: since I can "define" my entities twice as in typescript interfaces, I occasionally see two hints for the same entity name in the editor, since I create N descriptions for the same name basically. What's the best solution to deal with this? Attach an additional data structure to the document and "save" what I've already defined there so that I can double check if I'm adding another description for the same thing and avoid doing so?

Right now I'm wondering that maybe I did not understand what the ScopeProvider does. The `getScope` method only works on cross references if I'm not mistaken, so if I don't setup cross references correctly in my grammar I'm basically opting out from using the ScopeProvider to help me building the language tools. As long as I'm dealing with entities and function declarations (which happen to be defined globally at the root level only) I have no problem adding cross references for them to the grammar, since it's simple to add a `[Entity:ID]` or `[Function:ID]` on some grammar rule. But when it comes to variables/members/named values definition inside entities, it becomes a little bit trickier, since I can access them in various ways:

- directly (eg. `a`)
- fully qualified (eg. `A.a`)
- nested properties accessing for some kind of values (eg. `a.propertyAA.propertyAAA`, like objects in typescript)

And I can define them in various ways:

- simple way (eg. `a = 42`)
- destructuring (eg. `[a, ...] = [1,2,3]`)
- renamed destructuring (eg. `{ a3: a1 } = { a1: 42, a2: "hello" }` see this as destructuring + renaming)

And I have a "ValueRef" rule which is the following (I know it's not beautiful):

```
ValueRef:
    ref=ID ('.' access+=ID)*;
```

And I don't see how can I define a `NamedElement` to use as `[NamedElement:ID]` as you're advising in the [class member scoping guide](https://langium.org/guides/scoping/class-member/) for example.

**Reply by kristiannotari** [reply]
On top of that, I have some other blocks in my DSL which can implicitly define new entities if you put a name which is not used for already defined entities but I also want to have a cross reference there. The problem is that I can't have a cross reference mixed with a normal "ID" usage as in:
<img width="467" alt="image" src="https://user-images.githubusercontent.com/17885006/210218382-f9ec6dd8-787e-4999-9674-52dc9be96a7f.png">

**Reply by msujew** [reply]
> Not sure I've understood how following qualified names and member call scoping techniques could help me with the multiple entity definition

I probably misunderstood your problem.

> which is: since I can "define" my entities twice as in typescript interfaces, I occasionally see two hints for the same entity name in the editor, since I create N descriptions for the same name basically. What's the best solution to deal with this? 

Probably by overriding the `CompletionProvider` and filtering the results to prevent duplicates.

> The getScope method only works on cross references if I'm not mistaken, so if I don't setup cross references correctly in my grammar I'm basically opting out from using the ScopeProvider to help me building the language tools.

Yep, that's correct. The cross reference grammar syntax works closely together with the scope system to provide the whole cross referencing feature. Without it, Langium isn't able to know whether your `ID` is supposed to be just a plain id or a reference to another element.

> And I don't see how can I define a NamedElement to use as [NamedElement:ID] as you're advising in the [class member scoping guide](https://langium.org/guides/scoping/class-member/) for example.

What is your issue here exactly? You just need to put all possible elements which can be referenced in a union type and use that in your cross reference.

> On top of that, I have some other blocks in my DSL which can implicitly define new entities if you put a name which is not used for already defined entities but I also want to have a cross reference there. The problem is that I can't have a cross reference mixed with a normal "ID" usage as in

Yeah, that won't work. Although it might seem reasonable semantically, this doesn't work syntactically, since the parser can't identify whether you mean the `ID` that's supposed to reference a `Principal` element or the `ID` that's just a plain id.

One way you could accomplish this feature is by creating `AstNodeDescription` objects for those inferred elements, so that the scope can still resolve to a description, even though there's no real underlying element for that.

Alternatively, you can return some special scope object for those inferred elements, something like `ANY_SCOPE`, which gets special handling in the `Linker` implementation. Instead of creating a referencing error, it will just create an `undefined` `ref` value. That effectively simulates the `any` behavior in TypeScript.

**Reply by kristiannotari** [reply]
> Probably by overriding the CompletionProvider and filtering the results to prevent duplicates.

Thanks!

> What is your issue here exactly? You just need to put all possible elements which can be referenced in a union type and use that in your cross reference.

The problem being since I can have multiple types of access (direct, qualified, object nested properties, and so on), should I use the cross reference only for the "first" bit of the "potentially dot separated list of references for access" (eg. `a.b.c`) so that the first occurrence is something I actually have in scope? For example (I only have 1 level deep entity accessing values for fully qualified names, ever):
```
Reference: EntityDef | ValueDef; // yes I can reference entities

ValueRef:
    ref=[Reference:ID] ('.' access+=ID)*; // here I need [Reference:ID ('.' ID)?] really
```

From what I've understood I'd then supply the direct and qualified ways of accessing using a solution where I compute two descriptions for the same node, one with the brief name and one with the fully qualified name. For resolving object properties access I should rely on some autocompletion feature given a type system instead. Correct? 

> Yeah, that won't work. Although it might seem reasonable semantically, this doesn't work syntactically, since the parser can't identify whether you mean the ID that's supposed to reference a Principal element or the ID that's just a plain id.

So for this bit no ScopeProvider special implementation is needed I guess. Is the ScopeProvider useful for something else other than cross-reference resolution? Everything else I need for autocompletion stuff is based on specific work on validation/autocompletion IDE parts I guess. Correct?

> One way you could accomplish this feature is by creating AstNodeDescription objects for those inferred elements, so that the scope can still resolve to a description, even though there's no real underlying element for that.

Yeah I'm already doing so. Basically, when I compute scopes, I treat both the definition blocks and my other blocks where an ID can define a new entity as nodes from where I create description for such IDs (names).

**Reply by msujew** [reply]
> so that the first occurrence is something I actually have in scope?

At the bottom of the [member scoping guide](https://langium.org/guides/scoping/class-member/) there is a call graph which shows how the resolution is supposed to work. The first element is always an element of the local scope, while every nested call is an element of the previous call. That's why we transform the AST using the `{MemberCall.previous+=current}` action. To make it easier on the scoping algorithm to figure out what came before.

> From what I've understood I'd then supply the direct and qualified ways of accessing using a solution where I compute two descriptions for the same node, one with the brief name and one with the fully qualified name. For resolving object properties access I should rely on some autocompletion feature given a type system instead. Correct?

Note that qualified name access and nested object property access are somewhat in conflict. Looking at an expression like `a.b.c` is ambiguous from a parser perspective, as both are just `ID`s separated by `.`. The member call mechanism is generally better suited for a language like yours. The qualified name stuff is usually less powerful, and I wouldn't recommend using it in your case. So yes, combining member calls + a type system is the way to go.

> Is the ScopeProvider useful for something else other than cross-reference resolution?

Well, only in completion when the next (or current) element at the cursor position is a cross reference. Otherwise, yes, it's mainly used for the cross reference resolution mechanism.

**Comment by snarkipus** [comment]
Hello @kristiannotari ,

Here's some things I found helpful - I'll defer to the Langium team for deep technical advice, but this is what I use to try and understand what's going on under the hood.

DISCLAIMER: IMHO The best resource by far is the Langium implementation itself and I treat it as the 'reference implementation' for Langium. Anything provided below is probably '_almost correct'ish but occasionally not_'.

>I'd like not to bother you all by creating a discussion uniquely to guide me creating my DSL, but I wanted to get a clearer idea on how to structure things in Langium.

I have found the Langium team to be incredibly supportive and helpful. There's certainly a precedent for creating a dedicated thread for your DSL, so I wouldn't hesitate to do so ([my thread](https://github.com/langium/langium/discussions/609), [older thread](https://github.com/langium/langium/discussions/153)). I just try to return the favor by contributing to the project if and where I'm able.

One thing I found incredibly helpful for non-specific DSL design questions was the following book (based on Xtext):

[_Implementing Domain-Specific Languages with Xtext and Xtend - Second Edition_](https://www.packtpub.com/product/implementing-domain-specific-languages-with-xtext-and-xtend/9781786464965?_ga=2.24486655.4504868.1665617961-350183201.1665339294) (L. Bettini 2016).
 * [Xtext Source Code (Book)](https://github.com/PacktPublishing/Implementing-Domain-Specific-Languages-with-Xtext-and-Xtend)

Note: I attempted to follow along with the book using Langium and had relatively decent success - unfortunately, I've been too busy to finish the scoping portion. However, you're free to take a look as there's a pretty large amount of validation examples one would typically need.
* [Example DSL Repo](https://github.com/snarkipus/small-java)

>is there some way to visually grasp what's being built by the various "passes" from text to scopes and stuff? Like some diagrams populating based on cursor position in a DSL file or something like that? Cause that would help greatly to see what's being built behind the scene. Or is there another way to properly debug these "internals" without having to rely on editor hints to know what's available in scope?

I'm a picture person and not a SW dev by trade, so I drew this:
* [Document Life Cycle](https://app.excalidraw.com/s/7YrgtG5Ew9f/zXKm1VyNUa)

* For inspecting various data structures during different phases of the document build process, I added a test file like [this](https://github.com/snarkipus/langium-blocks/blob/main/src/test/block-playground-test.test.ts).

**Reply by kristiannotari** [reply]
Thanks for the advices! I have found them extremely supportive and fast at it too! Great job btw.

I'm not able to open your Document Life Cycle excalidraw link, even after registration, it keeps telling me to create a paid workspace (sorry, new to excalidraw, I'm probably doing something wrong on my side).

For your test case, the important bit is the console.log with some of the computed stuff for your document at [line 149](https://github.com/snarkipus/langium-blocks/blob/57e230fa38543697dc8c1e07fbbb2e03e19d19d3/src/test/block-playground-test.test.ts#L149), right? I'll give it a try to actually see how data structures are evolving and have quick feedback for quick changes and exploration stuff while trying langium. Thanks!

**Comment by snarkipus** [comment]
> I'm not able to open your Document Life Cycle excalidraw link, even after registration, it keeps telling me to create a paid workspace (sorry, new to excalidraw, I'm probably doing something wrong on my side).

How strange, I'll just embed it:


![document life cycle(2)](https://user-images.githubusercontent.com/3688989/210087326-7de7b496-4799-405b-8dfd-f1af024085fa.png)

> For your test case, the important bit is the console.log with some of the computed stuff for your document at [line 149](https://github.com/snarkipus/langium-blocks/blob/57e230fa38543697dc8c1e07fbbb2e03e19d19d3/src/test/block-playground-test.test.ts#L149), right?

That particular `console.log()` allows you to see the parsed document. What was more helpful for me was being able to peek at some of the various internal structures/services at various phases of the document lifecycle.

**Reply by kristiannotari** [reply]
Thanks a lot for your diagram, I'll try to associate the various steps with what I should do in my own DSL.

As of now I wanted to see the result of scope computation (for example) in a graphical way and I'm actually using your console.log (more or less) as a cli to debug files to see what I'm building with langium. Thanks!

**Reply by snarkipus** [reply]
>As of now I wanted to see the result of scope computation (for example) in a graphical way and I'm actually using your console.log (more or else) as a cli to debug files to see what I'm building with langium. 

I'm sure there's a more clever way of doing it, but that's the approach I landed on. Same challenge - working through scope computation.

**Reply by kristiannotari** [reply]
Yeah... you can see up [here](https://github.com/langium/langium/discussions/856#discussioncomment-4560504) I'm a little bit lost to be honest. Trying to picture what each step does to ease my work on building up validation for the language basically. I mean, I think I know how to write the scope computation, and I think to currently have a semi-working partial implementation of it, but I'm not sure what I should do with that.
